crates/ethrpc/src/buffered.rs

```diff
@@ -84,23 +92,48 @@ where
         tokio::task::spawn(batched_for_each(config, calls, move |batch| {
             let inner = inner.clone();
             async move {
-                let (mut requests, mut senders): (Vec<_>, Vec<_>) = batch
-                    .into_iter()
-                    .filter(|(_, _, sender)| !sender.is_canceled())
-                    .map(|(id, request, sender)| ((id, request), sender))
-                    .unzip();
+                let (mut requests, mut trace_ids, mut senders): (Vec<_>, Vec<_>, Vec<_>) =
+                    itertools::multiunzip(
+                        batch
+                            .into_iter()
+                            .filter(|(_, _, _, sender)| !sender.is_canceled())
+                            .map(|(id, request, trace_id, sender)| {
+                                ((id, request), trace_id, sender)
+                            }),
+                    );
                 match requests.len() {
                     0 => (),
                     1 => {
-                        let ((id, request), sender) = (requests.remove(0), senders.remove(0));
-                        let result = inner.send(id, request).await;
+                        let ((id, request), trace_id, sender) =
+                            (requests.remove(0), trace_ids.remove(0), senders.remove(0));
+                        let result = match (&request, trace_id) {
+                            (Call::MethodCall(_), Some(trace_id)) => {
+                                observe::request_id::set_task_local_storage(
+                                    trace_id,
+                                    inner.send(id, request),
+                                )
+                                .await
+                            }
+                            _ => inner.send(id, request).await,
+                        };
                         let _ = sender.send(result);
                     }
                     n => {
-                        let results = inner
-                            .send_batch(requests)
+                        let results = match build_rpc_metadata(&requests, &trace_ids) {
+                            Ok(metadata) => observe::request_id::set_task_local_storage(
+                                metadata,
+                                inner.send_batch(requests),
+                            )
                             .await
-                            .unwrap_or_else(|err| vec![Err(err); n]);
+                            .unwrap_or_else(|err| vec![Err(err); n]),
+                            // should never happen
```

Comment:
This is quite risky. If there is a bug in `build_rpc_metadata` we are unable to send any RPC requests (which will cause a complete outage).

I think the correct error handling here would be to still send the requests without local storage tracing context and simply log the error.

Reply:
Good point, fixed